contracts/libraries/ControllerV2.sol

```diff
@@ -0,0 +1,233 @@
+                require(success);
+            }
+        }
+    }
+
+    modifier wrapETH() {
+        if (address(this).balance > 0) _weth.deposit{value: address(this).balance}();
+        _;
+    }
+
+    function _executeCouponTrade(
+        address user,
+        address token,
+        Coupon[] memory couponsToMint,
+        Coupon[] memory couponsToBurn,
+        int256 interestThreshold
+    ) internal {
+        uint256 length = couponsToBurn.length + couponsToMint.length;
+        IController.Action[] memory actionList = new IController.Action[](length);
+        bytes[] memory paramsDataList = new bytes[](length);
+        address[] memory tokensToSettle = new address[](length);
+        IController.ERC20PermitParams[] memory erc20PermitParamsList;
+        IController.ERC721PermitParams[] memory erc721PermitParamsList;
+
+        uint256 amount;
+        address quote; // all quote is same
+
+        length = couponsToBurn.length;
+        for (uint256 i = 0; i < length; ++i) {
+            actionList[i] = IController.Action.TAKE;
+            IBookManager.BookKey memory key = _couponBuyMarkets[couponsToBurn[i].key.toId()];
+            quote = Currency.unwrap(key.quote);
+            amount += couponsToBurn[i].amount;
+            paramsDataList[i] = abi.encode(
+                IController.TakeOrderParams({
+                    id: key.toId(),
+                    limitPrice: type(uint256).max,
+                    quoteAmount: couponsToBurn[i].amount,
+                    hookData: ""
+                })
+            );
+        }
+        if (quote != address(0)) IERC20(quote).approve(address(_cloberController), amount);
+
+        length = couponsToMint.length;
+        for (uint256 i = 0; i < length; ++i) {
+            actionList[couponsToBurn.length + i] = IController.Action.SPEND;
+            IBookManager.BookKey memory key = _couponSellMarkets[couponsToMint[i].key.toId()];
+            amount = couponsToMint[i].amount;
+            paramsDataList[couponsToBurn.length + i] = abi.encode(
+                IController.SpendOrderParams({
+                    id: key.toId(),
+                    limitPrice: type(uint256).max,
+                    baseAmount: amount,
+                    hookData: ""
+                })
+            );
+            // key.base can't be Currency.NATIVE
+            IERC20(Currency.unwrap(key.base)).approve(address(_cloberController), amount);
+        }
+
+        if (interestThreshold > 0) {
+            IERC20(token).approve(address(_cloberController), uint256(interestThreshold));
+        }
+        _cloberController.execute(
+            actionList,
+            paramsDataList,
+            tokensToSettle,
+            erc20PermitParamsList,
+            erc721PermitParamsList,
+            uint64(block.timestamp)
+        );
+        if (interestThreshold < 0 && IERC20(token).balanceOf(address(this)) < uint256(-interestThreshold)) {
```

Comment:
We have to check the balance diff instead of the after state.

Reply:
Resolved.